File: 0x08-python-more_classes/4-rectangle.py
#!/usr/bin/python3
"""Define a class Rectangle."""


class Rectangle:
    """Represent a rectangle."""

    def __init__(self, width=0, height=0):
        """Initialize a new rectangle.

        Args:
            width (int): The width of the new rectangle
            height (int): The height of the  new rectanhle.
        """
        self.width = width
        self.height = height

    @property
    def width(self):
        """Get/set width of the rectangle."""
        return self.__width

    @width.setter
    def width(self, value):
        if not isinstance(value, int):
            raise TypeError("width must be an integer")
        if value < 0:
            raise ValueError("width must be >= 0")
        self.__width = value

    @property
    def height(self):
        """Get/set height of the rectangle."""
        return self.__height

    @height.setter
    def height(self, value):
        if not isinstance(value, int):
            raise TypeError("height must be an integer")
        if value < 0:
            raise ValueError("height must be >= 0")
        self.__height = value

    def area(self):
        """Return the area of the new rectangle."""
        return (self.__height * self.__width)

    def perimeter(self):
        """Return the parameter of the new rectangle."""
        if self.__width == 0 or self.__height == 0:
            return 0
        return ((self.__height + self.__width) * 2)

    def __str__(self):
        """Return printable representation of the rectangle.

        Represent the rectangle with the character #.
        """
        if self.__width == 0 or self.__height == 0:
            return("")

        rect = []
        for column in range(self.__height):
            for row in range(self.__width):
                rect.append("#")
            if column != self.__height - 1:
                rect.append("\n")
        return ("".join(rect))

    def __repr__(self):
        """Return the string representation of the Rectangle."""
        rect = "Rectangle(" + str(self.__width)
        rect += ", " + str(self.__height) + ")"
        return (rect)

File: 0x13-javascript_objects_scopes_closures/4-rectangle.js
#!/usr/bin/node
module.exports = class Rectangle {
  constructor (w, h) {
    if (w > 0 && h > 0) { [this.width, this.height] = [w, h]; }
  }

  print () {
    // Prints the rectangle using the character X
    for (let r = 0; r < this.height; r++) {
      let row = '';
      for (let c = 0; c < this.width; c++) row += 'X';
      console.log(row);
    }
  }

  rotate () {
    // Exchanges the width and the height of the rectangle.
    const tmp = this.width;
    this.width = this.height;
    this.height = tmp;
  }

  double () {
    // Multiples the width and the height of the rectangle by 2.
    this.width *= 2;
    this.height *= 2;
  }
};
